packages/cosmic-swingset/lib/ag-solo/vats/vat-mailbox.js

```diff
@@ -0,0 +1,48 @@
+import harden from '@agoric/harden';
+import { makeRegistrar } from '@agoric/registrar';
+import { E } from '@agoric/eventual-send';
+import { assert } from '@agoric/assert';
+
+function build(_log) {
+  const mailboxRegistry = makeRegistrar();
+
+  const mailboxAdmin = harden({
+    makeMailbox: (mailboxPrefix, purse, action = 'deposit') => {
+      assert.typeof(action, 'string');
+      const mailbox = harden({
+        receivePayment: payment => {
+          switch (action) {
+            case 'deposit': {
+              return E(purse).deposit(payment);
+            }
+            default: {
+              throw new Error(`action ${action} is not implemented`);
+            }
+          }
+        },
+      });
+      const key = mailboxRegistry.register(mailboxPrefix, mailbox);
+      return key;
+    },
+    sendPayment: (mailboxKey, payment) => {
+      const mailbox = mailboxRegistry.get(mailboxKey);
+      return mailbox.receivePayment(payment);
+    },
+    // We deliberately do not expose the entire registry so that the
+    // only way someone can send a payment using a mailbox is if they
+    // have obtained the key from the creator of a mailbox.
```

Comment:
The registrar abstraction was designed only to hold capabilities (like issuers) that could be widely held. I do see that you're not using it this way, and that you might be using it in a way that does in fact uphold the security properties you need. But why use the registry at all? It would be simpler to just use a weakStore directly.

Reply:
Using a weakStore directly is not simpler, because it would not manage the key namespace in the way I require. A weakStore needs unique keys and throws an error if a key exists, thus revealing information about what is in the weakStore. So, the weakStore has fewer security properties in this case, not more.

The registry abstraction manages the key namespace in the right way for this purpose. It always creates a new unique key, and the creation of a new mapping does not leak any information about keys that already exist. 

Attenuating access to the registry is within ocap practice. :) Even though we built the registry thinking that everyone would be able to get all the values, there's no reason why we can't wrap the registry and limit access.

Reply:
You are using the registry in exactly two places: line 24 where you call `register` and line 28 where you call `get`.

You'd replace line 24 with

```js
const key = harden({});
weakStore.init(key, mailbox);
return key;
```
No possibility of collision because you just made a new key.

Line 28 would remain essentially the same.

What namespace? What is this prefix for? No one touches the registry other than those two places so who would see this prefix anyway?

If the registry had been intended to be used this way, I would have given it a very different review.

This pattern of using the weakStore is essentially the sealer/unsealer pattern, an oldie but goodie of ocap programming going back to 1973

Reply:
Ah, part of the requirements for this task is that the key be a string. It cannot be a handle because we want people to able to communicate it through other channels and potentially through the UI.

Reply:
TL;DR Nevermind.

Offline, @katelynsills clarified that the key is a string that is not intended to be secret. If Alice registers a mailbox, this design assumes it is fine with her if anyone sends her money, even if they came to have knowledge of the key by illicit means.

This raises the separate question of why the public registry is not a good place for alice to put a capability to send her money. But that seems to be an issue with the design of the public registry, not with this PR.

Reply:
I updated the comment to make it clearer. Let me know if you think it can be improved.